### PR TITLE
feat: redirect log

### PR DESCRIPTION
**Pull request by nguyenhoangthuan99**
Fix https://github.com/janhq/cortex.cpp/issues/1054
Summary changed:
- Add `file_logger` implementation as `cortex-cpp` did
- Create an interface `SetFileLogger(int max_log_lines, const std::string& log_path)` so that when server in `cortex-cpp` load engines, it can enable file logger by calling this function

### CODE REVIEW

File: base/cortex-common/enginei.h [hunk 1]
```diff
@@ -36,7 +36,8 @@ class EngineI {
     }
     return false;
   }
-  virtual void SetFileLogger() = 0;
+  virtual void SetFileLogger(int max_log_lines,
```
Comment (vansangpfiev) [inline base/cortex-common/enginei.h:39]: Please move this function to the end of the class

File: src/llama_engine.h [hunk 2]
```diff
@@ -1,14 +1,15 @@
...
+#include "llama.h"
 #include "llama_server_context.h"
 #include "trantor/utils/ConcurrentTaskQueue.h"
-#include "llama.h"
-#include <trantor/utils/AsyncFileLogger.h>
 
 constexpr char log_base_name[] = "logs/cortex";
 constexpr char log_folder[] = "logs";
-constexpr size_t max_log_file_size = 20000000; // ~20mb
+constexpr size_t max_log_file_size = 20000000;  // ~20mb
```
Comment (dan-homebrew) [inline src/llama_engine.h:12]: Question: is this settable via `.cortexrc`? (and is it per-engine specific?)

**Review by dan-homebrew — APPROVED** [review]

File: src/llama_engine.h (hunk 2 above)
Comment (nguyenhoangthuan99) [reply src/llama_engine.h:12]: New logger don't need that config, I'll remove max_log_file_size

File: base/cortex-common/enginei.h (hunk 1 above)
Comment (nguyenhoangthuan99) [reply base/cortex-common/enginei.h:39]: I'll remove it

**Comment by dan-homebrew** [comment]
@nguyenhoangthuan99 I've approved this PR to unblock you. 

However, can I verify my understanding of this PR, and the separation of concerns between the main `cortex` and `llama.cpp-engine`

- `cortex.llamacpp` has its own file logger (?)
- `cortex.cpp` uses `SetFileLogger` to pass the path to the llama.cpp engine, which then implements its own file logging

I have the following questions:

- If we run multiple engines, will there be lock problems on the log file (or is it one process with cortex.cpp?)
- If each engine implements its own logging, will there be a standardized logging format? (e.g. from Trantor verbose, with timestamps)

**Comment by nguyenhoangthuan99** [comment]
> @nguyenhoangthuan99 I've approved this PR to unblock you.
> 
> However, can I verify my understanding of this PR, and the separation of concerns between the main `cortex` and `llama.cpp-engine`
> 
> * `cortex.llamacpp` has its own file logger (?)
> * `cortex.cpp` uses `SetFileLogger` to pass the path to the llama.cpp engine, which then implements its own file logging
> 
> I have the following questions:
> 
> * If we run multiple engines, will there be lock problems on the log file (or is it one process with cortex.cpp?)
> * If each engine implements its own logging, will there be a standardized logging format? (e.g. from Trantor verbose, with timestamps)

- When we run multiple engines, all engines still run in the same process, but they can in different thread, so the lock problem won't appear.
- The File logger implementation is copied from cortex.cpp so every format still the same, this is one limitation that we separate into many repo, some part of code we cannot reuse and we have to copy it manually

**Review by vansangpfiev — APPROVED** [review]

**Pull request merged**